include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -3878,6 +3879,64 @@ struct OrtApi {
                   _In_z_ const char* config_key, _Out_ char* config_value, _Inout_ size_t* size);
 
   /// @}
+  /// \name OrtKernelInfo
+  /// Custom operator APIs.
+  /// @{
+
+  /** \brief Get the session logger from ::OrtKernelInfo.
+   *
+   * Used in the CreateKernel callback of an OrtCustomOp to get a logger that can be used to log
+   * messages.
+   *
+   * \param[in] info An instance of ::OrtKernelInfo.
+   * \param[out] logger Pointer set to the session's ::OrtLogger. Do not free.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   * \since Version 1.15
+   */
+  ORT_API2_STATUS(KernelInfo_GetLogger, _In_ const OrtKernelInfo* info, _Outptr_ const OrtLogger** logger);
+
+  /// @}
+  /// \name OrtLogger
+  /// Custom operator APIs.
+  /// @{
+
+  /** \brief Logs a message at the given severity level using the provided ::OrtLogger.
+   *
+   * Only messages with a severity level equal or greater than the ::OrtLogger's logging severity level
+   * are logged. Use OrtApi::Logger_GetLoggingSeverityLevel to get the ::OrtLogger's logging severity
+   * level.
+   *
+   * Can be used in custom operators to log messages with the logger retrieved via OrtApi::KernelInfo_GetLogger.
+   *
+   * \param[in] logger The ::OrtLogger instance.
+   * \param[in] log_severity_level The message's severity level.
+   * \param[in] message The message to log.
+   * \param[in] file_path The filepath of the file in which the message is logged. Usually the value of __FILE__.
+   * \param[in] line_number The file line number in which the message is logged. Usually the value of __LINE__.
+   * \param[in] func_name The name of the function in which the message is logged. Usually the value of __FUNCTION__.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   * \since Version 1.15
+   */
+  ORT_API2_STATUS(Logger_LogMessage, _In_ const OrtLogger* logger, OrtLoggingLevel log_severity_level,
```

Comment:
>const OrtLogger* logge [](http://example.com/codeflow?start=42&length=22)

Is logging a mutating operation? How is this represented downstream?

Reply:
Logging is currently a const operation in internal ort code.

https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/common/logging/logging.cc#L156
https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/common/logging/capture.h#L31

Also, the logger that is stored in the kernel context or in the execution provider is a const pointer/reference.

Aside: We could decouple the constness of the OrtLogger from that of the internal onnxruntime::logging::Logger by making OrtLogger a wrapper class that contains a ref/pointer to a onnxruntime::logging::Logger. We are then free to apply our own semantics for logging. This comes at the cost of an extra memory allocation and an additional API call to release the OrtLogger, but it lets us be more flexible.